File: veo/apps/api/src/veo/db/models/keywords.py
"""Naver keyword intelligence.

The three Naver data families never share a column:

* ``NAVER_SEARCH_AD``  — official absolute monthly search counts, clicks, CTR, competition.
* ``NAVER_DATALAB``    — a *relative* interest index. Not a search count, ever.
* ``CALCULATED``       — VEO's own arithmetic, always labelled with its formula version.

Zero, missing, provider-suppressed and below-threshold are distinct facts and are stored
as distinct facts. A suppressed value is never written as 0.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from veo.db.base import (
    Base,
    ImmutableMixin,
    JsonArray,
    JsonObject,
    OrganizationScopedMixin,
    TimestampMixin,
    json_column,
    uuid_pk,
)


class KeywordQuery(Base, OrganizationScopedMixin, ImmutableMixin):
    """One lookup request, kept so a figure can always be traced to when it was asked."""

    __tablename__ = "keyword_queries"
    __table_args__ = (Index("ix_keyword_queries_org_created", "organization_id", "created_at"),)

    id: Mapped[uuid.UUID] = uuid_pk()
    project_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    job_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("jobs.id", ondelete="SET NULL"), nullable=True
    )
    surface: Mapped[str] = mapped_column(String(16), nullable=False, default="CONSOLE")
    original_keyword: Mapped[str] = mapped_column(String(255), nullable=False)
    normalized_keyword: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    locale: Mapped[str] = mapped_column(String(16), nullable=False, default="ko-KR")
    requested_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    provider_state: Mapped[str] = mapped_column(String(32), nullable=False, default="ENABLED")
    error_code: Mapped[str | None] = mapped_column(String(48), nullable=True)


class KeywordMetric(Base, OrganizationScopedMixin, ImmutableMixin):
    """Official Naver Search Ad figures for one keyword at one point in time.

    Each numeric column has a paired ``*_quality`` column recording whether the value is
    exact, rounded, a range, suppressed by the provider, below its reporting threshold, or
    simply missing. Reading a count without reading its quality is a bug.
    """

    __tablename__ = "keyword_metrics"
    __table_args__ = (
        UniqueConstraint("keyword_query_id", "normalized_keyword"),
        Index("ix_keyword_metrics_keyword_collected", "normalized_keyword", "collected_at"),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    keyword_query_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("keyword_queries.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    normalized_keyword: Mapped[str] = mapped_column(String(255), nullable=False)

    source: Mapped[str] = mapped_column(String(32), nullable=False, default="NAVER_SEARCH_AD")
    api_version: Mapped[str | None] = mapped_column(String(32), nullable=True)
    collected_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    source_period: Mapped[str | None] = mapped_column(String(32), nullable=True)
    raw_response_hash: Mapped[str | None] = mapped_column(String(64), nullable=True)
    was_cache_hit: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    monthly_pc_searches: Mapped[int | None] = mapped_column(Integer, nullable=True)
    monthly_pc_searches_quality: Mapped[str] = mapped_column(
        String(32), nullable=False, default="MISSING"
    )
    monthly_mobile_searches: Mapped[int | None] = mapped_column(Integer, nullable=True)
    monthly_mobile_searches_quality: Mapped[str] = mapped_column(
        String(32), nullable=False, default="MISSING"
    )
    monthly_total_searches: Mapped[int | None] = mapped_column(
        Integer, nullable=True, comment="CALCULATED from the two device figures, not provided."
    )

    avg_pc_clicks: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_mobile_clicks: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_pc_ctr: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_mobile_ctr: Mapped[float | None] = mapped_column(Float, nullable=True)

    competition_index: Mapped[float | None] = mapped_column(Float, nullable=True)
    competition_label: Mapped[str | None] = mapped_column(String(32), nullable=True)
    ad_depth: Mapped[int | None] = mapped_column(
        Integer, nullable=True, comment="Ad slots observed. Advertising competition only."
    )

    provider_raw: Mapped[JsonObject] = json_column()
    partial_reason: Mapped[str | None] = mapped_column(Text, nullable=True)


class RelatedKeyword(Base, OrganizationScopedMixin, ImmutableMixin):
    __tablename__ = "related_keywords"
    __table_args__ = (
        Index("ix_related_keywords_query_rank", "keyword_query_id", "source_rank"),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    keyword_query_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("keyword_queries.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    seed_keyword: Mapped[str] = mapped_column(String(255), nullable=False)
    related_keyword: Mapped[str] = mapped_column(String(255), nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False, default="NAVER_SEARCH_AD")
    source_rank: Mapped[int | None] = mapped_column(Integer, nullable=True)
    monthly_total_searches: Mapped[int | None] = mapped_column(Integer, nullable=True)
    value_quality: Mapped[str] = mapped_column(String(32), nullable=False, default="MISSING")


class KeywordTrend(Base, OrganizationScopedMixin, ImmutableMixin):
    """DataLab relative interest. Stored apart from counts so it can never be mistaken for one."""

    __tablename__ = "keyword_trends"
    __table_args__ = (
        UniqueConstraint("keyword_query_id", "normalized_keyword", "period_start", "device"),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    keyword_query_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("keyword_queries.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    normalized_keyword: Mapped[str] = mapped_column(String(255), nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False, default="NAVER_DATALAB")
    period_start: Mapped[date] = mapped_column(Date, nullable=False)
    period_end: Mapped[date] = mapped_column(Date, nullable=False)
    time_unit: Mapped[str] = mapped_column(String(16), nullable=False, default="month")
    device: Mapped[str] = mapped_column(String(16), nullable=False, default="ALL")
    relative_index: Mapped[float] = mapped_column(
        Float,
        nullable=False,
        comment="Relative interest, 0-100 within the requested window. NOT a search count.",
    )
    index_basis_note_ko: Mapped[str | None] = mapped_column(Text, nullable=True)
    collected_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class KeywordList(Base, OrganizationScopedMixin, TimestampMixin):
    __tablename__ = "keyword_lists"
    __table_args__ = (UniqueConstraint("project_id", "name"),)

    id: Mapped[uuid.UUID] = uuid_pk()
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    keywords: Mapped[JsonArray] = json_column()


class KeywordOpportunity(Base, OrganizationScopedMixin, ImmutableMixin):
    """VEO's own opportunity score.

    Explicitly a VEO calculation, never a provider figure. Every component is stored
    separately so a customer can see exactly what drove the number, and the formula
    version travels with it.
    """

    __tablename__ = "keyword_opportunities"
    __table_args__ = (
        UniqueConstraint("keyword_query_id", "normalized_keyword", "formula_version"),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    keyword_query_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("keyword_queries.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    normalized_keyword: Mapped[str] = mapped_column(String(255), nullable=False)
    formula_version: Mapped[str] = mapped_column(String(32), nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False, default="CALCULATED")

    demand: Mapped[float | None] = mapped_column(Float, nullable=True)
    trend: Mapped[float | None] = mapped_column(Float, nullable=True)
    intent_fit: Mapped[float | None] = mapped_column(Float, nullable=True)
    competition_inverse: Mapped[float | None] = mapped_column(Float, nullable=True)
    content_gap: Mapped[float | None] = mapped_column(Float, nullable=True)
    confidence: Mapped[float] = mapped_column(Float, nullable=False)
    opportunity_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    calculation_trace: Mapped[JsonObject] = json_column()
    missing_components: Mapped[JsonArray] = json_column()

File: veo/apps/web/src/lib/keywords.ts
import 'server-only';

import { callConsoleApi, type ConsoleOutcome } from '@/lib/console-api';

/**
 * 네이버 키워드 — 엔진에서 읽어 오는 쪽.
 *
 * 이 파일에서 절대 하지 않는 것: **숫자 하나로 줄이기.**
 *
 * 엔진은 모든 수치를 `{value, quality, source, note_ko}` 로 준다. 그 네 개가 함께
 * 있어야 "3,200회" 와 "네이버가 값을 숨김" 과 "너무 적어 집계 안 됨" 이 구분된다.
 * 화면 편하자고 `number | null` 로 접으면 그 구분이 여기서 사라지고, 뒤에서 되살릴
 * 방법이 없다.
 *
 * 그리고 **검색량과 관심도 지수를 같은 타입으로 만들지 않는다.** 절대량(SearchAd)과
 * 상대지수(DataLab)는 단위가 다르다. 타입이 같으면 언젠가 같은 표에 들어간다.
 */

/** 값이 왜 그 모양인지. 0 과 "자료 없음" 은 다른 사실이다. */
export type ValueQuality =
  | 'EXACT'
  | 'ROUNDED'
  | 'RANGE'
  | 'SUPPRESSED_BY_PROVIDER'
  | 'BELOW_PROVIDER_THRESHOLD'
  | 'MISSING';

export type DataSource =
  | 'NAVER_SEARCH_AD'
  | 'NAVER_DATALAB'
  | 'NAVER_SEARCH_API'
  | 'CALCULATED'
  | 'VEO_INTERNAL';

/** 절대 건수. 맨 숫자로 오는 일이 없다. */
export interface MeasuredCount {
  readonly value: number | null;
  readonly quality: ValueQuality;
  readonly source: DataSource;
  readonly note_ko: string;
  readonly upper_bound_exclusive: number | null;
}

export interface MeasuredRatio {
  readonly value: number | null;
  readonly quality: ValueQuality;
  readonly source: DataSource;
  readonly note_ko: string;
}

export interface KeywordMetrics {
  readonly source: DataSource;
  readonly api_version: string | null;
  readonly collected_at: string;
  readonly age_seconds: number;
  readonly source_period: string | null;
  readonly cache_hit: boolean;

  readonly monthly_pc_searches: MeasuredCount;
  readonly monthly_mobile_searches: MeasuredCount;
  readonly monthly_total_searches: MeasuredCount;

  readonly avg_pc_clicks: MeasuredRatio;
  readonly avg_mobile_clicks: MeasuredRatio;
  readonly avg_pc_ctr: MeasuredRatio;
  readonly avg_mobile_ctr: MeasuredRatio;

  readonly competition_label: string | null;
  /** 네이버는 지수를 주지 않는다. 늘 `null` 이고, VEO 가 숫자를 지어내지 않는다. */
  readonly competition_index: number | null;
  readonly ad_depth: number | null;
  readonly partial_reason_ko: string | null;
}

export interface TrendPoint {
  readonly period_start: string;
  readonly relative_index: number;
}

/**
 * DataLab 관심도 추세.
 *
 * **검색 횟수가 아니다.** 필드 이름에 `count`·`searches` 가 없는 것은 의도된 것이고,
 * 화면에서도 "건" 을 붙이면 안 된다.
 */
export interface KeywordTrend {
  readonly source: DataSource;
  readonly unit: string;
  readonly note_ko: string;
  readonly time_unit: string;
  readonly device: string;
  readonly period_start: string;
  readonly period_end: string;
  readonly points: readonly TrendPoint[];
}

export interface OpportunityComponent {
  readonly name: string;
  readonly weight: number;
  readonly value: number | null;
  readonly contribution: number | null;
  readonly source: DataSource;
  readonly note_ko: string;
  readonly unavailable_reason_ko: string | null;
}

/** VEO 자체 산식. 네이버가 준 값이 아니라는 사실이 화면에 남아야 한다. */
export interface Opportunity {
  readonly source: DataSource;
  readonly formula_version: string;
  readonly score: number | null;
  readonly coverage: number;
  readonly freshness: number;
  readonly confidence: number;
  readonly components: readonly OpportunityComponent[];
  readonly missing_components: readonly string[];
  readonly disclosure_ko: string;
  readonly unavailable_reason_ko: string | null;
}

export interface RelatedKeyword {
  readonly seed_keyword: string;
  readonly related_keyword: string;
  readonly source: DataSource;
  readonly source_rank: number | null;
  readonly monthly_total_searches: MeasuredCount;
}

export interface KeywordSnapshot {
  readonly original_keyword: string;
  readonly normalized_keyword: string;
  readonly metrics: KeywordMetrics | null;
  readonly trend: KeywordTrend | null;
  readonly opportunity: Opportunity | null;
  readonly related: readonly RelatedKeyword[];
}

/** 쓸 수 없는 공급자도 이유와 함께 돌려준다 — 빼면 "여기 있는 게 전부" 로 읽힌다. */
export interface ProviderBadge {
  readonly provider: string;
  readonly state: string;
  readonly reason_ko: string;
}

export interface KeywordLookup {
  readonly query_id: string | null;
  readonly project_id: string | null;
  readonly requested_at: string;
  readonly locale: string;
  readonly providers: readonly ProviderBadge[];
  readonly keywords: readonly KeywordSnapshot[];
  readonly notices_ko: readonly string[];
}

export interface RecentKeywords {
  readonly title_ko: string;
  readonly source: DataSource;
  readonly window_hours: number;
  readonly period_start: string;
  readonly period_end: string;
  readonly entries: readonly {
    readonly normalized_keyword: string;
    readonly lookup_count: number;
    readonly last_requested_at: string;
  }[];
  readonly disclosure_ko?: string;
}

/** 조회는 네이버를 실제로 부른다. 목록 조회와 시간 감각이 다르다. */
export async function lookUpKeywords(
  keywords: readonly string[],
): Promise<ConsoleOutcome<KeywordLookup>> {
  return callConsoleApi('/api/keywords/lookups', {
    method: 'POST',
    body: { keywords },
    timeoutMs: 60_000,
  });
}

export async function readRecentKeywords(): Promise<ConsoleOutcome<RecentKeywords>> {
  return callConsoleApi('/api/keywords/recent');
}
